### PR TITLE
change: Enable default model fn for cpu and gpu

### PR DESCRIPTION
**Pull request by saimidu**
*Issue #, if available:*

*Description of changes:*
Add a default model function for CPU/GPU based inference, instead of raising a `NotImplementedError`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: buildspec.yml [hunk 1]
```diff
@@ -39,7 +39,7 @@ phases:
       - GENERIC_TAG="$FRAMEWORK_VERSION-pytorch-$BUILD_ID"
       - DLC_CPU_TAG="$FRAMEWORK_VERSION-dlc-cpu-$BUILD_ID"
       - DLC_GPU_TAG="$FRAMEWORK_VERSION-dlc-gpu-$BUILD_ID"
-      - DLC_EIA_TAG="$FRAMEWORK_VERSION-dlc-eia-$BUILD_ID"
+      - DLC_EIA_TAG="$EIA_FRAMEWORK_VERSION-dlc-eia-$BUILD_ID"
```
Comment (davidthomas426) [inline buildspec.yml:42]: This is an unrelated bug fix, right?

Comment (davidthomas426) [reply buildspec.yml:42]: By the way, is EIA_FRAMEWORK_VERSION actually correct? Later in the review, a comment mentions EIA using version 1.5.1, but this file has the variable set to '1.3.1'.

File: test/integration/sagemaker/test_default_inference.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        model_data=model_data,
+        role="SageMakerRole",
+        predictor_cls=RealTimePredictor if not accelerator_type else PyTorchPredictor,
+        entry_point=mnist_script,
+        image=image_uri,
+        sagemaker_session=sagemaker_session,
+    )
+    with timeout_and_delete_endpoint(endpoint_name, sagemaker_session, minutes=30):
+        # Use accelerator type to differentiate EI vs. CPU and GPU. Don't use processor value
+        if accelerator_type is not None:
```
Comment (davidthomas426) [inline test/integration/sagemaker/test_default_inference.py:84]: Isn't this branch pointless? If it is None, then the code will do the same thing but use the default value of `accelerator_type` parameter, which is most likely `None` (though I didn't check)

File: test/integration/sagemaker/test_default_inference.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        else:
+            predictor = pytorch.deploy(
+                initial_instance_count=1, instance_type=instance_type, endpoint_name=endpoint_name
+            )
+
+        if accelerator_type:
+            batch_size = 100
+            data = np.random.rand(batch_size, 1, 28, 28).astype(np.float32)
+            output = predictor.predict(data)
+            assert output.shape == (batch_size, 10)
```
Comment (davidthomas426) [inline test/integration/sagemaker/test_default_inference.py:100]: Why is the data and assert different for EIA vs CPU/GPU? Aren't they both the same or similar models?

Comment (davidthomas426) [reply test/integration/sagemaker/test_default_inference.py:100]: And why is the output of `predictor.predict(...)` handled differently in the two cases? In one case, `.shape` is looked up directly on it, while the other case decodes it and passes it to `json.loads`. What's going on here?

Comment (davidthomas426) [reply test/integration/sagemaker/test_default_inference.py:100]: Ok...looking further in the review, I found my answer. EIA path uses mnist with default input/output handlers, while CPU/GPU path uses resnet18 with transform_fn defined. Got it!

File: test/resources/mnist/default_model_eia/code/mnist.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+logger.addHandler(logging.StreamHandler(sys.stdout))
+
+
+def predict_fn(input_data, model):
+    logger.info('Performing EIA inference with Torch JIT context with input of size {}'.format(input_data.shape))
+    # With EI, client instance should be CPU for cost-efficiency.
+    # Sub-graphs with unsupported arguments run locally. Server runs with CUDA
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
```
Comment (davidthomas426) [inline test/resources/mnist/default_model_eia/code/mnist.py:29]: Actually, in order for the EIA Client Libraray to work, the device has to be set to 'cpu', regardless of if CUDA is available. If device is set to 'cuda', I can't remember what exactly happens with the below `{'target_device': 'eia:0'}` code, but I don't think it uses the EI accelerator.

File: buildspec.yml (hunk 1 above)
Comment (saimidu) [reply buildspec.yml:42]: Yes, that's right. It is an unrelated bug fix, but both 1.3.1 and 1.5.1 are untestable right now, since they use `mms` instead of `torchserve`. This toolkit is now built around `torchserve` which makes it necessary to publish a PyTorch DLC image for EIA based on `torchserve`.

File: test/integration/sagemaker/test_default_inference.py (hunk 2 above)
Comment (saimidu) [reply test/integration/sagemaker/test_default_inference.py:84]: You are right. The default for `accelerator_type` is None. So the if and else conditions can both be collapsed into one single implementation.

File: test/integration/sagemaker/test_default_inference.py (hunk 3 above)
Comment (saimidu) [reply test/integration/sagemaker/test_default_inference.py:100]: Since I haven't been able to test EIA images with the new toolkit, and the test is skipped by default, I copied over the test prediction request from the `test_mnist.py` test. The success or failure of this EIA test will only be known once we start running it on an image that is capable of using the new toolkit.

File: src/sagemaker_pytorch_serving_container/default_pytorch_inference_handler.py [hunk 5]
```diff
@@ -49,10 +48,14 @@ def default_model_fn(self, model_dir):
...
             return torch.jit.load(model_path, map_location=torch.device('cpu'))
         else:
-            raise NotImplementedError(textwrap.dedent("""
-            Please provide a model_fn implementation.
-            See documentation for model_fn at https://github.com/aws/sagemaker-python-sdk
-            """))
+            device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+            model_path = os.path.join(model_dir, DEFAULT_MODEL_FILENAME)
+            if not os.path.exists(model_path):
+                raise FileNotFoundError("Failed to load model with default model_fn: missing file {}."
```
[the remark this answers is not in the document]
Comment (saimidu) [reply src/sagemaker_pytorch_serving_container/default_pytorch_inference_handler.py]: Please take a look at the updated model file search.
